language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/dependencygraph/DependencyGraph.scala

```diff
@@ -47,7 +51,24 @@ private final case class LFDependencyGraph(private val util: lf.LFUtil)
             recDeps ++ choiceDeps,
             collectDepError = true))
     }
-    Graph.cyclicDependencies(internalNodes = typeDeclNodes, roots = templateNodes)
+
+    def matchesRoots(qualName: Ref.Identifier): Boolean =
+      util.roots.exists(_.findFirstIn(qualName.qualifiedName.qualifiedName).isDefined)
+    // scala-2.13-M4: _.matches(qualName.qualifiedName.qualifiedName)
+
+    val interesting =
+      if (util.roots.isEmpty)
+        templateNodes
+      else {
+        val res = templateNodes.filter(pair => matchesRoots(pair._1))
+        logger.info(s"""Filtered template names to match regex set:
+             |${util.roots.mkString(", ")}.
+             |Considering ${res.size} templates.
+           """.stripMargin)
+        res
+      }
```

Comment:
The filtering done here is too deeply nested; instead, the `EnvironmentInterface` itself should be filtered closer to `CodeGen`. Then probably `Util` doesn't need to have the regex.